### PR TITLE
Accomodate null character after authentication message on some implementations of scp1

### PR DESCRIPTION
**Pull request by TomCos**
Accidental comment, see next one :)

### CODE REVIEW

**Comment by TomCos** [comment]
So hear me out, I've been working heavily on some Napalm/Netmiko projects and some legacy Cisco devices put the null characters at the end of, after their authentication message, instead of before.  My re-write is small, but here is what it looks like.  I put print statements of msg after each receive:

b'Authentication success\\r\\n\\x00'
b'\\x00'
b'\\x00'

It continues with no issue, but without checking the end for the null character, the exception is thrown.

File: scp.py [hunk 1]
```diff
@@ -355,9 +355,9 @@ def _recv_confirm(self):
         except SocketTimeout:
             raise SCPException('Timeout waiting for scp response')
         # slice off the first byte, so this compare will work in py2 and py3
-        if msg and msg[0:1] == b'\x00':
+        if msg and type(msg) == bytes and (repr(str(msg).startswith('\x00')) or repr(str(msg).endswith('\x00'))):
             return
-        elif msg and msg[0:1] == b'\x01':
+        elif msg and type(msg) == bytes and repr(str(msg).startswith('\x01')):
```
Comment (remram44) [inline scp.py]: Why check the type? Does paramiko's `recv()` ever return something else?

Comment (TomCos) [reply scp.py]: Ya I don't think it does, but I was trying to be accommodating because I converted the bytes type to a string.

Comment (remram44) [reply scp.py]: You can check for a null byte without converting

Comment (TomCos) [reply scp.py]: More than happy to accept a better implementation.  I was having trouble checking null byte at the end.

**Comment by remram44** [comment]
So I am not sure what this would mean in practice. It is probable that no null byte ever appears in the message anyway, so looking for it at any position doesn't matter, but I don't know. Do you know if this is something that is supported by other clients (for example OpenSSH's scp(1))?

File: scp.py (hunk 1 above)
Comment (remram44) [reply scp.py]: I don't understand why you are calling `repr()` either?

**Comment by TomCos** [comment]
I agree, I'd rather not break all the working implementations out there, but I am confused on something.  Maybe it's my ignorance, but you are looking for the reply to start with a null byte character in your code, but in your comment, you say that no null byte might show up, which would throw an exception in current state.  My assumption is, you're very familiar with how low level SCP works at this point, and there is something in the implementation of it that said you should expect that before data is sent?  For what I've been using your package for (which is awesome btw), this is the only other scenario I've hit where it "doesn't work," so I'm assuming I might not be the only person hitting the bug, which in this case, is actually server side (the Cisco switch), not your module (client).  I did test your implementation, much better than mine!

**Comment by remram44** [comment]
Sorry if I wasn't clear!

The protocol as I understand it should look like this: `<0/1/2><message>`, but if we know there is never a 0/1/2 byte in the message, we can just check whether such a byte occurs anywhere. That would still catch the "normal" situation, but would also catch the `<cisco-auth-message><NUL><message>` that you seem to face.

**Comment by TomCos** [comment]
Hey @remram44 just wanted to see if we can to any conclusion on this one? 😃

**Comment by remram44** [comment]
It would be easier to move forward if you could point out an available implementation that behaves this way and that I could test against. Otherwise I don't know if it's reasonable to twist the protocol this way.

**Comment by TomCos** [comment]
Just want to follow up again, as I'm not sure how I can help further.  I thought the log I posted above would show the issue fairly well, but if we're looking for more data, I just need some guidance.  Like I said, I don't have experience in the scp1 protocol at a low enough level to make a call where the null character should and shouldn't show up.  The comment you made above sounds like to me to check the entire line for the null character, instead of the start, but like I said, I feel that might be too big of a change and could cause issues.  Checking the beginning and the end seems like a middle ground, based on the fact that I don't see anyone else complaining about this issue, and I've shown above that it can occur at the end, based on server implementation.  Let me know what we need to move forward and I will gladly participate if I can.

**Comment by TomCos** [comment]
Any decision on how to handle this one?

**Comment by TomCos** [comment]
@remram44 Following up again

**Comment by remram44** [comment]
I don't think this is the right way to handle this. It would be good to know how other implementations deal with this. I am worried this first message is unrelated to the commands sent but just the authentication, and we're really ignoring the actual code when using that one instead.

I think having a way for me to test on an actual server, or a pointer to some other SCP implementation which deals with this, is necessary to go forward.

**Comment by TomCos** [comment]
I did test it by modifying the scp package in earlier comments and it did work, so safe to say it's probably just an odd implementation.

What I'm trying to figure out is, why did we decide to check the beginning for the b'\x01' character?  Is there something in the spec that specifically says it?  My thinking is, Cisco did follow the spec and the spec says b'\x01' has to be at the end.  Normally it's the only thing seen in most implementations (or no one has ever complained), which is why this came up.

I'm still having trouble understanding why checking the beginning and the end holds any extra risk?

**Comment by remram44** [comment]
As far as I can see, yes, this is how OpenSSH implements it: [read](https://github.com/openssh/openssh-portable/blob/4d28fa78abce2890e136281950633fae2066cc29/scp.c#L1560-L1596) and [write](https://github.com/openssh/openssh-portable/blob/4d28fa78abce2890e136281950633fae2066cc29/scp.c#L1608-L1631).

**Comment by TomCos** [comment]
I don't really have any experience with C, so it's tough to read through the code.  From the excerpt you posted though, yes it looks like there should not even be a message after the binary 0, let alone in front.  I think that all makes sense.  I'm having trouble wrapping my head around why it works with a terminal, but any terminal is likely using SCP is actually using SFTP if I'm not mistaken, so maybe that's the answer.  Based on that, agreed we can probably close this one?

**Comment by remram44** [comment]
I'm starting to wonder whether this might be a bug in Paramiko actually. Can you execute commands ok with Paramiko against that server? Do you perhaps see this "Authentication success" message in the output of commands you execute?

**Comment by TomCos** [comment]
No not in every command, just after connecting it's in the data of the first command run.

**Comment by TomCos** [comment]
@remram44 any idea why when doing it through just the linux SCP command we don't see the issue?  I've tried on both Linux and OSX and the transfer goes through.  "Authentication success" still prints, but the transfer continues anyways.  I can't really read the C code very well, but do you see anything in the read implementation that would explain this?

**Comment by TomCos** [comment]
Been about a year on this one.  Any chance of moving forward?

**Comment by remram44** [comment]
It's hard for me to evaluate without access to the affected server, but it does look like something that should be fixed in Paramiko if any command (not just SCP ones) show the wrong output.

**Pull request closed without merging**